### ISSUE TITLE
Implement vsync handling in the Voodoo code

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#3976]
## Overview

I've been suspecting that most (or perhaps all?) Voodoo DOS games enforced vsync, while in our Voodoo implementation the vsync handling is commented out.

This results in going _way_ above 60 FPS when multithreaded Voodoo emulation is enabled, which consumes a lot of computing power as the Voodoo code is quite resource-intensive.

I asked Aaron Giles, the original author of the Voodoo code about this, and it looks like I was right. We'll need to re-enable the vsync paths and let the games set the vsync bit at the hardware level. Probably all Voodoo games set this bit, but we'll need to test it.

See the transcript below for further details.

## Transcript

**Me:** did the Voodoo 1 enforce vsync by default in DOS games? note we're only interested in DOS games. my understanding (which might be totally wrong) is that the Voodoo pretty much only supported 640x480 @ 60 Hz in DOS games, and that's it. it couldn't support higher resolutions because it only had 4 MB on onboard RAM.

**Aaron:** There's nothing in the hardware that forced VSYNC, but it was common to do so. The swapBufferCMD had a flag to sync to vsync, and I'm sure Glide leveraged that. Not sure how you integrated that behavior into DOSBox, but I'd start there and see if the games in question are setting the low bit.

![image](https://github.com/user-attachments/assets/ecdecac5-0e62-4765-91bf-d47497f29114)

**Me:** so in your experience, most games turned vsync on, then?
if a game set that bit, was there any way to override that "from the outside"? e.g. by setting certain env vars in DOS, etc.

**Aaron:** This is at the hardware level, so no. Now, Glide might check environment variables and alter the value passed to the command, but if you log the values passed to swapbufferCMD they should match what the hardware would see.

## References

The above reference is from page 55 of the Voodoo specs:
http://www.o3one.org/hwdocs/video/voodoo_graphics.pdf

**Comment by weirddan455** [comment]
I agree this would be something good to implement.  It won't solve high CPU usage on its own though.  The worker threads are busy waiting since #3952 because there was a lot of overhead in waiting and waking the threads otherwise.

We probably need some method of batching the work into larger units before dispatching to the worker threads.  I measured when I made that PR and currently each thread is getting sent work about every 100 microseconds or so.

**Comment by johnnovak** [comment]
Yeah, but if busy waiting gets you from say 150 to 170 FPS (in some games), but the game enforces vsync at 60 Hz, we'd be fine with the non-busy-wait 150 FPS algo as we'll be capped to 60 FPS anyway 😄 

Of course, if it's possible to keep the busy waiting performance while not burning up the CPU, even better.

**Comment by weirddan455** [comment]
>  Yeah, but if busy waiting gets you from say 150 to 170 FPS (in some games), but the game enforces vsync at 60 Hz, we'd be fine with the non-busy-wait 150 FPS algo as we'll be capped to 60 FPS anyway

I wouldn't have made the PR if it was only 10-20% gains.  @interloper98 and @FeralChild64 both reported 40-50% improvement in Quake 1/2 and it pushed performance above that 60fps mark which was not possible on without busy wait.  I was also getting 300%+ gains in Descent 2 where it would sometimes dip under 60fps with no busy wait.

**Comment by johnnovak** [comment]
> > Yeah, but if busy waiting gets you from say 150 to 170 FPS (in some games), but the game enforces vsync at 60 Hz, we'd be fine with the non-busy-wait 150 FPS algo as we'll be capped to 60 FPS anyway
> 
> I wouldn't have made the PR if it was only 10-20% gains. @interloper98 and @FeralChild64 both reported 40-50% improvement in Quake 1/2 and it pushed performance above that 60fps mark which was not possible on without busy wait. I was also getting 300%+ gains in Descent 2 where it would sometimes dip under 60fps with no busy wait.

Cool cool. Yeah, forgive me, I didn't remember the exact figures and didn't bother to look the older posts up 😄 

Then it's a matter of making that busy wait less expensive or implement some queuing like you said.

**Comment by weirddan455** [comment]
Just to update this ticket for a future reader: We're no longer busy-waiting in Voodoo since #4220 so high CPU usage and noisy fans are less of a concern now.

We may still want proper vsync handling to reduce tearing though.
